File: java/Samples/Biometrics/VB/AbisSampleVB/Settings/SettingsTab.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms

Partial Public Class SettingsTab
	Inherits TabPageContentBase
	Implements IPageController
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		TabName = "Settings"
	End Sub

#End Region

#Region "Private fields"

	Private _pages As New List(Of SettingsPageBase)()
	Private _currentPage As SettingsPageBase = Nothing

#End Region

#Region "Public properties"

	Public Shadows Property TabController() As ITabController Implements IPageController.TabController
		Get
			Return MyBase.TabController
		End Get
		Set(ByVal value As ITabController)
			MyBase.TabController = value
		End Set
	End Property

#End Region

#Region "Public methods"

	Public Overrides Sub OnTabAdded()
		If listViewPages.SelectedItems.Count = 0 Then
			NavigateToStartPage()
		End If

		MyBase.OnTabAdded()
	End Sub

	Public Overrides Sub SetParams(ByVal ParamArray parameters() As Object)
		If parameters IsNot Nothing AndAlso parameters.Length = 1 Then
			Try
				Dim param As String = parameters(0).ToString()
				Dim type As Type = type.GetType(param)
				NavigateToPage(type, TabController.Client)
			Catch
			End Try
		End If

		MyBase.SetParams(parameters)
	End Sub

	Public Sub NavigateToPage(ByVal pageType As Type, ByVal navigationParam As Object, ByVal ParamArray arguments() As Object) Implements IPageController.NavigateToPage
		If pageType Is Nothing OrElse (Not pageType.IsSubclassOf(GetType(SettingsPageBase))) Then
			Throw New ArgumentException("pageType")
		End If
		If _currentPage IsNot Nothing AndAlso _currentPage.GetType() Is pageType Then
			Return
		End If

		If _currentPage IsNot Nothing Then

			_currentPage.OnNavigatingFrom()
			_currentPage.NavigationParam = Nothing
			_currentPage = Nothing
			panelPage.Controls.Clear()
		End If

		If arguments Is Nothing OrElse arguments.Length = 0 Then
			arguments = New Object() {navigationParam}
		End If
		Dim page As SettingsPageBase = _pages.FirstOrDefault(Function(x) x.GetType() Is pageType)
		If page Is Nothing Then
			page = CType(Activator.CreateInstance(pageType), SettingsPageBase)
			page.PageController = Me
			_pages.Add(page)
		End If

		_currentPage = page
		panelPage.Controls.Add(page)

		page.NavigationParam = navigationParam
		page.OnNavigatedTo(arguments)

		For Each item As ListViewItem In listViewPages.Items
			If item.Tag Is pageType Then
				item.Selected = True
				Exit For
			End If
		Next item
	End Sub

	Public Sub NavigateToStartPage() Implements IPageController.NavigateToStartPage
		NavigateToPage(GetType(GeneralSettingsPage), TabController.Client)
	End Sub

#End Region

#Region "Private form events"

	Private Sub ListViewPagesSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles listViewPages.SelectedIndexChanged
		Dim selected = listViewPages.SelectedItems
		Dim selectedItem As ListViewItem = Nothing
		If selected.Count > 0 Then
			selectedItem = listViewPages.SelectedItems(0)
			NavigateToPage(CType(selectedItem.Tag, Type), TabController.Client)
		End If
	End Sub

	Private Sub BtnOkClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOk.Click
		If _currentPage IsNot Nothing Then
			_currentPage.SaveSettings()
		End If
		SettingsManager.SaveSettings(TabController.Client)
		TabController.CloseTab(Me)
	End Sub

	Private Sub BtnCancelClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnCancel.Click
		SettingsManager.LoadSettings(TabController.Client)
		TabController.CloseTab(Me)
	End Sub

	Private Sub BtnDefaultClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnDefault.Click
		If _currentPage IsNot Nothing Then
			_currentPage.DefaultSettings()
		End If
	End Sub

	Private Sub SettingsTabLoad(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			listViewPages.Items.Add(New ListViewItem("General") With {.Tag = GetType(GeneralSettingsPage)})
			If LicensingTools.CanCreateFingerTemplate(TabController().Client.LocalOperations) Then
				listViewPages.Items.Add(New ListViewItem("Fingers") With {.Tag = GetType(FingersSettingsPage)})
			End If
			If LicensingTools.CanCreateFaceTemplate(TabController().Client.LocalOperations) Then
				listViewPages.Items.Add(New ListViewItem("Faces") With {.Tag = GetType(FacesSettingsPage)})
			End If
			If LicensingTools.CanCreateIrisTemplate(TabController().Client.LocalOperations) Then
				listViewPages.Items.Add(New ListViewItem("Irises") With {.Tag = GetType(IrisesSettingsPage)})
			End If
			If LicensingTools.CanCreatePalmTemplate(TabController().Client.LocalOperations) Then
				listViewPages.Items.Add(New ListViewItem("Palms") With {.Tag = GetType(PalmsSettingsPage)})
			End If
			If LicensingTools.CanCreateVoiceTemplate(TabController().Client.LocalOperations) Then
				listViewPages.Items.Add(New ListViewItem("Voices") With {.Tag = GetType(VoicesSettingsPage)})
			End If
		End If
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/Settings/SettingsTab.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;

namespace Neurotec.Samples
{
	public partial class SettingsTab : TabPageContentBase, IPageController
	{
		#region Public constructor

		public SettingsTab()
		{
			InitializeComponent();
			TabName = "Settings";
		}

		#endregion

		#region Private fields

		private List<SettingsPageBase> _pages = new List<SettingsPageBase>();
		private SettingsPageBase _currentPage = null;

		#endregion

		#region Public methods

		public override void OnTabAdded()
		{
			if (listViewPages.SelectedItems.Count == 0)
				NavigateToStartPage();

			base.OnTabAdded();
		}

		public override void SetParams(params object[] parameters)
		{
			if (parameters != null && parameters.Length == 1)
			{
				try
				{
					string param = parameters[0].ToString();
					Type type = Type.GetType(param);
					NavigateToPage(type, TabController.Client);
				}
				catch
				{
				}
			}

			base.SetParams(parameters);
		}

		public void NavigateToPage(Type pageType, object navigationParam, params object[] arguments)
		{
			if (pageType == null || !pageType.IsSubclassOf(typeof(SettingsPageBase))) throw new ArgumentException("pageType");
			if (_currentPage != null && _currentPage.GetType() == pageType) return;

			if (_currentPage != null)
			{

				_currentPage.OnNavigatingFrom();
				_currentPage.NavigationParam = null;
				_currentPage = null;
				panelPage.Controls.Clear();
			}

			if (arguments == null || arguments.Length == 0) arguments = new[] { navigationParam };
			SettingsPageBase page = _pages.FirstOrDefault(x => x.GetType() == pageType);
			if (page == null)
			{
				page = (SettingsPageBase)Activator.CreateInstance(pageType);
				page.PageController = this;
				_pages.Add(page);
			}

			_currentPage = page;
			panelPage.Controls.Add(page);

			page.NavigationParam = navigationParam;
			page.OnNavigatedTo(arguments);

			foreach (ListViewItem item in listViewPages.Items)
			{
				if (item.Tag == pageType)
				{
					item.Selected = true;
					break;
				}
			}
		}

		public void NavigateToStartPage()
		{
			NavigateToPage(typeof(GeneralSettingsPage), TabController.Client);
		}

		#endregion

		#region Private form events

		private void ListViewPagesSelectedIndexChanged(object sender, EventArgs e)
		{
			var selected = listViewPages.SelectedItems;
			ListViewItem selectedItem = null;
			if (selected.Count > 0)
			{
				selectedItem = listViewPages.SelectedItems[0];
				NavigateToPage((Type)selectedItem.Tag, TabController.Client);
			}
		}

		private void BtnOkClick(object sender, EventArgs e)
		{
			if (_currentPage != null) _currentPage.SaveSettings();
			SettingsManager.SaveSettings(TabController.Client);
			TabController.CloseTab(this);
		}

		private void BtnCancelClick(object sender, EventArgs e)
		{
			SettingsManager.LoadSettings(TabController.Client);
			TabController.CloseTab(this);
		}

		private void BtnDefaultClick(object sender, EventArgs e)
		{
			if (_currentPage != null)
				_currentPage.DefaultSettings();
		}

		private void SettingsTabLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				listViewPages.Items.Add(new ListViewItem("General") { Tag = typeof(GeneralSettingsPage) });
				if (LicensingTools.CanCreateFingerTemplate(TabController.Client.LocalOperations))
					listViewPages.Items.Add(new ListViewItem("Fingers") { Tag = typeof(FingersSettingsPage) });
				if (LicensingTools.CanCreateFaceTemplate(TabController.Client.LocalOperations))
					listViewPages.Items.Add(new ListViewItem("Faces") { Tag = typeof(FacesSettingsPage) });
				if (LicensingTools.CanCreateIrisTemplate(TabController.Client.LocalOperations))
					listViewPages.Items.Add(new ListViewItem("Irises") { Tag = typeof(IrisesSettingsPage) });
				if (LicensingTools.CanCreatePalmTemplate(TabController.Client.LocalOperations))
					listViewPages.Items.Add(new ListViewItem("Palms") { Tag = typeof(PalmsSettingsPage) });
				if (LicensingTools.CanCreateVoiceTemplate(TabController.Client.LocalOperations))
					listViewPages.Items.Add(new ListViewItem("Voices") { Tag = typeof(VoicesSettingsPage) });
			}
		}

		#endregion
	}
}
